### PR TITLE
[Feature] Add callback to reduce lr on plateaus

### PR DESCRIPTION
**Pull request by inafergra**
This PR adds a new callback to reduce the learning rate when a metric does not improve after a given number of epochs, in the spirit of `torch`'s [ReduceLROnPlateau](https://docs.pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html). I developed it for a project and thought it would be worth including, it has worked very well to help QNNs converge

### CODE REVIEW

**Review by chMoussa — APPROVED** [review]
LGTM @inafergra

File: perceptrain/callbacks/callback.py [hunk 1]
```diff
@@ -666,6 +666,134 @@ def run_callback(self, trainer: Any, config: TrainConfig, writer: BaseWriter) ->
...
+    def run_callback(self, trainer: Any, config: TrainConfig, writer: BaseWriter) -> None:
+        """
+        Reduces the learning rate when the loss reaches a plateau.
+
+        Args:
+            trainer (Any): The training object.
+            config (TrainConfig): The configuration object.
+            writer (BaseWriter): The writer object for logging.
+        """
+        if not self.reached_minimum:
```
Comment (debrevitatevitae) [inline perceptrain/callbacks/callback.py:763]: Change name of variable for clarity

```suggestion
        if not self.reached_minimum_lr:
```

File: perceptrain/callbacks/callback.py [hunk 2]
```diff
@@ -666,6 +666,134 @@ def run_callback(self, trainer: Any, config: TrainConfig, writer: BaseWriter) ->
...
+                            logger.info(
+                                f"Learning rate has reached the minimum learning rate {self.min_lr}"
+                            )
+                    else:
+                        param_group["lr"] = new_lr
+                        if self.verbose:
+                            logger.info(
+                                f"Loss has reached a plateau, reducing learning rate to {new_lr}"
+                            )
+                self.counter = 0
```
Comment (debrevitatevitae) [inline perceptrain/callbacks/callback.py:794]: For clarity, change the control flow to happy-path first
```suggestion
        if self.reached_minimum_lr:
            pass
        else:
            current_value = trainer.opt_result.metrics.get(self.monitor)
            if current_value is None:
                raise ValueError(
                    f"Metric '{self.monitor}' is not available in the trainer's metrics."
                )

            if (self.mode == "min" and current_value + self.threshold < self.best_value) or (
                self.mode == "max" and current_value - self.threshold > self.best_value
            ):
                self.best_value = current_value
                self.counter = 0
            else:
                self.counter += 1

            if self.counter >= self.patience:
                for param_group in trainer.optimizer.param_groups:
                    new_lr = param_group["lr"] * self.gamma
                    if new_lr < self.min_lr:
                        param_group["lr"] = self.min_lr
                        self.reached_minimum = True
                        if self.verbose:
                            logger.info(
                                f"Learning rate has reached the minimum learning rate {self.min_lr}"
                            )
                    else:
                        param_group["lr"] = new_lr
                        if self.verbose:
                            logger.info(
                                f"Loss has reached a plateau, reducing learning rate to {new_lr}"
                            )
                self.counter = 0
```

File: tests/ml_tools/test_callbacks.py [hunk 3]
```diff
@@ -197,6 +198,48 @@ def test_lr_scheduler_cosine_annealing(trainer: Trainer) -> None:
     assert math.isclose(new_lr, expected_lr, rel_tol=1e-6)
 
 
+def test_lr_scheduler_reduce_on_plateau(trainer: Trainer) -> None:
```
Comment (debrevitatevitae) [inline tests/ml_tools/test_callbacks.py:201]: The test logic looks really good.

About the format though, it would be clearer if you could split it into some fixtures. E.g.
- fixt1: returns you the callback instance
- fixt2(fixt1, parameter=num_updates): updates the callback and returns you the LR

**Review by debrevitatevitae — COMMENTED** [review]
Thanks @inafergra for the nice work.

Just minor comments.

**Review by debrevitatevitae — APPROVED** [review]

File: tests/ml_tools/test_callbacks.py (hunk 3 above)
Comment (inafergra) [reply tests/ml_tools/test_callbacks.py]: Hey @debrevitatevitae thanks a lot for the comments. I added the fixtures as suggested, but I couldn't find a way around using fixture requests which doesn't make it too clear I think, let me you had another thing in mind

Comment (debrevitatevitae) [reply tests/ml_tools/test_callbacks.py]: Looks clear to me, thanks @inafergra !
